py/desispec/io/emlinefit.py

```diff
@@ -0,0 +1,567 @@
+        from desitarget.targetmask import desi_mask as mask
+    elif len(dtkeys) == 1:
+        dtkey = dtkeys[0]
+        if dtkey == "CMX_TARGET":
+            from desitarget.cmx.cmx_targetmask import cmx_mask as mask
+        if dtkey == "SV1_DESI_TARGET":
+            from desitarget.sv1.sv1_targetmask import desi_mask as mask
+        if dtkey == "SV2_DESI_TARGET":
+            from desitarget.sv2.sv2_targetmask import desi_mask as mask
+        if dtkey == "SV3_DESI_TARGET":
+            from desitarget.sv3.sv3_targetmask import desi_mask as mask
+    else:
+        log.info(
+            "found {}>1 matching keys: {}; 0 or 1 key expected; exiting".format(
+                len(dtkeys), dtkeys
+            )
+        )
+        sys.exit(1)
+    allowed_bitnames = mask.names()
+    # AR read + select the targetids
+    d = fitsio.read(redrockfn, columns=["TARGETID", dtkey], ext="FIBERMAP")
+    sel = np.zeros(len(d), dtype=bool)
+    for bitname in bitnames.split(","):
+        if bitname not in allowed_bitnames:
+            log.info(
+                "{} not in allowed bitnames for {} ({}; exiting)".format(
+                    bitname, dtkey, allowed_bitnames
+                )
+            )
+            sys.exit(1)
+        sel |= (d[dtkey] & mask[bitname]) > 0
+    targetids = d["TARGETID"][sel]
+    log.info("selecting {} targets with {} in {}".format(sel.sum(), bitnames, dtkey))
+    return targetids
+
+
+def get_rf_em_waves(emname):
+    """
+    Returns the rest-frame, vacuum, wavelengths.
+    Args:
+        emname, from: "OII", "HDELTA", "HGAMMA", "HBETA", "OIII", "HALPHA" (string)
+
+    Returns:
+        rf_em_waves: rest-frame wavelength(s) (np array)
+
+    Notes:
+        For OII and OIII returns a two-elements array; one-element array otherwise.
+        https://github.com/desihub/fastspecfit/blob/60393296e0cc466858f70a5d021d02693eff9375/py/fastspecfit/data/emlines.ecsv
+    """
+    if emname == "OII":
+        rf_em_waves = np.array([3727.092, 3729.874])
+    if emname == "OIII":
+        rf_em_waves = np.array([4960.295, 5008.239])
+    if emname == "HALPHA":
+        rf_em_waves = np.array([6564.613])
+    if emname == "HBETA":
+        rf_em_waves = np.array([4862.683])
+    if emname == "HGAMMA":
+        rf_em_waves = np.array([4341.684])
+    if emname == "HDELTA":
+        rf_em_waves = np.array([4102.892])
+    return rf_em_waves
+
+
+def emlines_gaussfit(
+    emname,
+    targetid,
+    zspec,
+    waves,
+    fluxes,
+    ivars,
+    rf_fit_hw=40,
+    min_rf_fit_hw=20,
+    rf_cont_w=200,
+    p0_sigma=2.5,
+    p0_flux=10,
+    p0_share=0.58,
+    min_sigma=1e-5,
+    max_sigma=10.,
+    min_flux=1e-5,
+    max_flux=1e9,
+    min_share=1e-1,
+    max_share=1,
+    log=get_logger(),
+):
+    """
+    Fits the [OII] doublet line profile with 2 related Gaussians.
+
+    Args:
+        emname: "OII" or "OIII" or "HALPHA", "HBETA", "HGAMMA", "HDELTA" (string)
+        targetid: targetid (int)
```

Comment:
targetid appears to be used only for passing through into the output, and thus feels like it shouldn't be a required input.  i.e. from a core algorithm standpoint, fitting gaussians to data shouldn't require a targetid as input.

Reply:
done (nice catch).